kyo-core/shared/src/main/scala/kyo/Fiber.scala

```diff
@@ -294,6 +296,189 @@ object Fiber extends FiberPlatformSpecific:
             }
         }
 
+    /** Concurrently executes effects and collects their successful results.
+      *
+      * WARNING: Executes all computations in parallel without bounds. Use with caution on large sequences to avoid resource exhaustion.
+      *
+      * Executes all effects concurrently and returns successful results in completion order. If all computations fail, the last encountered
+      * error is propagated. The operation completes when all effects have either succeeded or failed.
+      *
+      * @tparam Ctx
+      *   Context requirements
+      * @param seq
+      *   Sequence of effects to execute
+      * @return
+      *   Fiber containing successful results as a Chunk
+      */
+    inline def gather[E, A: Flat, Ctx](seq: Seq[A < (Abort[E] & Async & Ctx)])(
+        using
+        frame: Frame,
+        safepoint: Safepoint
+    ): Fiber[E, Chunk[A]] < (IO & Ctx) =
+        val total = seq.size
+        _gather(total)(total, seq)
+    end gather
+
+    /** Concurrently executes effects and collects up to `max` successful results.
+      *
+      * WARNING: Executes all computations in parallel without bounds. Use with caution on large sequences to avoid resource exhaustion.
+      *
+      * Similar to `gather`, but completes early once the specified number of `max` successful results is reached. If not enough successes
+      * occur and all remaining computations fail, the last encountered error is propagated.
+      *
+      * @param max
+      *   Maximum number of successful results to collect
+      * @param seq
+      *   Sequence of effects to execute
+      * @return
+      *   Fiber containing successful results as a Chunk (size <= max)
+      */
+    inline def gather[E, A: Flat, Ctx](max: Int)(seq: Seq[A < (Abort[E] & Async & Ctx)])(
+        using
+        frame: Frame,
+        safepoint: Safepoint
+    ): Fiber[E, Chunk[A]] < (IO & Ctx) =
+        _gather(max)(seq.size, seq)
+
+    private[kyo] def _gather[E, A: Flat, Ctx](max: Int)(total: Int, seq: Seq[A < (Abort[E] & Async & Ctx)])(
+        using
+        boundary: Boundary[Ctx, IO & Abort[E]],
+        frame: Frame,
+        safepoint: Safepoint
+    ): Fiber[E, Chunk[A]] < (IO & Ctx) =
+        if total == 0 || max <= 0 then Fiber.success(Chunk.empty)
+        else
+            IO.Unsafe {
+                class State extends IOPromise[E, Chunk[A]] with Function2[Int, Result[E, A], Unit]:
+                    val results = new Array[AnyRef](max)
+
+                    // Helper array to store original indices to maintain ordering
+                    // Initialized to Int.MaxValue to handle partial results
+                    val indices = new Array[Int](max)
+                    Arrays.fill(indices, Int.MaxValue)
```

Comment:
It's definitely more heavyweight since it allocates thunks and functions + has to dispatch them. `Arrays.fill` is a simple for loop amenable to auto vectorization. I imagine JIT compilers might even replace the method with intrinsics.